arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/IQueue.kt

```diff
@@ -47,52 +51,68 @@ internal class IQueue<A> internal constructor(
   fun exists(p: (A) -> Boolean): Boolean =
     listIn.any(p) || listOut.any(p)
 
-  fun length(): Int = listIn.size + listOut.size
-
   fun enqueue(elem: A): IQueue<A> =
-    IQueue(cons(elem, listIn), listOut)
+    IQueue(cons(elem, listIn), listOut, size + 1)
 
   fun enqueue(elems: Iterable<A>): IQueue<A> =
-    IQueue(elems.reversed() + listIn, listOut)
+    IQueue(elems.reversed() + listIn, listOut, size + elems.size())
 
   fun prepend(elem: A): IQueue<A> =
-    IQueue(listIn, cons(elem, listOut))
+    IQueue(listIn, cons(elem, listOut), size + 1)
 
   fun dequeue(): Pair<A, IQueue<A>> =
     when {
       listOut.isEmpty() && listIn.isNotEmpty() -> {
         val rev = listIn.reversed()
-        Pair(rev.first(), IQueue(emptyList(), rev.tail()))
+        Pair(rev.first(), IQueue(emptyList(), rev.tail(), size - 1))
       }
-      listOut.isNotEmpty() -> Pair(listOut.first(), IQueue(listIn, listOut.tail()))
+      listOut.isNotEmpty() -> Pair(listOut.first(), IQueue(listIn, listOut.tail(), size - 1))
       else -> throw NoSuchElementException("dequeue on empty queue")
     }
 
   fun drop(n: Int): IQueue<A> = when {
-    listOut.isEmpty() && listIn.isNotEmpty() -> IQueue(emptyList(), listIn.reversed().drop(n))
-    listOut.isNotEmpty() -> IQueue(listIn, listOut.drop(n))
-    else -> throw NoSuchElementException("dequeue on empty queue")
+    listOut.isEmpty() && listIn.isNotEmpty() -> IQueue(emptyList(), listIn.reversed().drop(n), size - n.coerceAtLeast(0))
+    listOut.isNotEmpty() -> IQueue(listIn, listOut.drop(n), size - n.coerceAtLeast(0))
+    else -> empty()
   }
 
   fun dequeueOrNull(): Pair<A, IQueue<A>>? =
     if (isEmpty()) null
     else dequeue()
 
-  fun filter(p: (A) -> Boolean): IQueue<A> =
-    IQueue(listIn.filter(p), listOut.filter(p))
+  fun filter(p: (A) -> Boolean): IQueue<A> {
+    val newIn = listIn.filter(p)
+    val newOut = listOut.filter(p)
+    return IQueue(newIn, newOut, newIn.size + newOut.size)
+  }
+
+  fun filterNot(p: (A) -> Boolean): IQueue<A> {
+    val newIn = listIn.filterNot(p)
+    val newOut = listOut.filterNot(p)
+    return IQueue(newIn, newOut, newIn.size + newOut.size)
+  }
 
-  fun filterNot(p: (A) -> Boolean): IQueue<A> =
-    IQueue(listIn.filterNot(p), listOut.filterNot(p))
+  override fun equals(other: Any?): Boolean =
```

Comment:
Is there a reason for IQueue to not be a data class? Given its a class defined by its properties

Reply:
No good point!
It can definitely be a data class, I'd still overwrite the `toString` to the new one tho.

Reply:
That'd make sense to me.